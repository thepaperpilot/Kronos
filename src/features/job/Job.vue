<template>
    <span
        v-if="unref(visibility) !== Visibility.None"
        :style="[
            {
                visibility: unref(visibility) === Visibility.Hidden ? 'hidden' : undefined,
                '--posx': unref(imageFocus).x,
                '--posy': unref(imageFocus).y,
                '--progress': `${(1 - unref(levelProgress)) * 100}%`,
                '--foreground': unref(color)
            },
            unref(style) ?? {}
        ]"
        class="feature dontMerge job"
        :class="{
            selected,
            active: selected || unref(timeLoopActive),
            animating
        }"
        @click="openJob"
    >
        <img :src="unref(image)" />
        <div class="job-contents">
            <div class="job-resource" v-for="(resource, index) in resourceArray" :key="index">
                {{ displayResource(resource) }} {{ resource.displayName }}
            </div>
            <div class="job-title">
                <h2>{{ name }}</h2>
                <span style="margin-bottom: 2px">Lv. {{ formatWhole(level.value) }}</span>
                <span style="flex-grow: 1" />
                <ModifierInfo
                    v-if="modifierInfo"
                    :display="modifierInfo"
                    :name="name"
                    v-bind="modifierModalAttrs"
                />
            </div>
        </div>
        <div v-if="selected && unref(currentQuip)" class="job-quip">"{{ unref(currentQuip) }}"</div>
        <div class="job-progress-container">
            <div class="job-progress"></div>
        </div>
        <Tooltip :direction="Direction.Left" display="Toggle Time Loop" class="job-loop-toggle">
            <button
                class="material-icons"
                @click.stop="timeLoopActive.value = !unref(timeLoopActive)"
                v-if="finishedFirstChapter"
                :class="{
                    active: unref(timeLoopActive)
                }"
                :disabled="!hasTimeSlotAvailable && !unref(timeLoopActive)"
            >
                all_inclusive
            </button>
        </Tooltip>
        <Node :id="id" />
    </span>
</template>

<script lang="tsx">
import "components/common/features.css";
import Node from "components/Node.vue";
import { main } from "data/projEntry";
import { CoercableComponent, StyleValue, Visibility } from "features/feature";
import ModifierInfo from "features/ModifierInfo.vue";
import { displayResource, Resource } from "features/resources/resource";
import Tooltip from "features/tooltips/Tooltip.vue";
import { Persistent } from "game/persistence";
import player from "game/player";
import { formatWhole } from "util/bignum";
import { Direction } from "util/common";
import { processedPropType, unwrapRef } from "util/vue";
import {
    computed,
    ComputedRef,
    defineComponent,
    onUnmounted,
    PropType,
    ref,
    Ref,
    toRefs,
    unref,
    watch,
    watchEffect
} from "vue";

export default defineComponent({
    props: {
        visibility: {
            type: processedPropType<Visibility>(Number),
            required: true
        },
        style: processedPropType<StyleValue>(String, Object, Array),
        classes: processedPropType<Record<string, boolean>>(Object),
        xp: {
            type: Object as PropType<Resource>,
            required: true
        },
        level: {
            type: Object as PropType<Resource>,
            required: true
        },
        levelProgress: {
            type: Object as PropType<Ref<number>>,
            required: true
        },
        timeLoopActive: {
            type: Object as PropType<Persistent<boolean>>,
            required: true
        },
        name: {
            type: String,
            required: true
        },
        color: {
            type: processedPropType<string>(String),
            required: true
        },
        image: {
            type: processedPropType<string>(String),
            required: true
        },
        imageFocus: {
            type: processedPropType<{ x: string; y: string }>(Object),
            required: true
        },
        resource: [Object, Array] as PropType<Resource | Resource[]>,
        layerID: {
            type: String,
            required: true
        },
        currentQuip: {
            type: Object as PropType<Ref<string | null>>,
            required: true
        },
        randomQuips: {
            type: processedPropType<string[]>(Array),
            required: true
        },
        modifierInfo: processedPropType<CoercableComponent>(String, Object, Function),
        id: {
            type: String,
            required: true
        },
        modifierModalAttrs: Object as PropType<Record<string, unknown>>
    },
    components: {
        Node,
        ModifierInfo,
        Tooltip
    },
    setup(props) {
        const { layerID, currentQuip, randomQuips, resource } = toRefs(props);

        const selected = computed(() => player.tabs[1] == layerID.value);
        const finishedFirstChapter: ComputedRef<boolean> = computed(() => main.chapter.value > 1);
        const hasTimeSlotAvailable: ComputedRef<boolean> = main.hasTimeSlotAvailable;

        function openJob() {
            player.tabs.splice(1, 1, layerID.value);
        }

        let quipTimer: number | undefined = undefined;
        watchEffect(() => {
            clearInterval(quipTimer);
            const quips = unwrapRef(randomQuips);
            if (currentQuip.value) {
                quipTimer = setTimeout(() => {
                    currentQuip.value = null;
                }, 15000);
            } else {
                quipTimer = setTimeout(() => {
                    currentQuip.value = quips[Math.floor(Math.random() * quips.length)];
                }, Math.random() * 15000 + 15000);
            }
        });
        onUnmounted(() => clearInterval(quipTimer));

        const animating = ref<boolean>(false);
        watch(selected, () => {
            animating.value = true;
            setTimeout(() => (animating.value = false), 250);
        });

        const resourceArray = computed(() => {
            if ("displayName" in resource) {
                return [resource] as unknown as Resource[];
            }
            const currResource = unref(resource);
            if (Array.isArray(currResource)) {
                return currResource as unknown as Resource[];
            }
            return [];
        });

        return {
            selected,
            animating,
            finishedFirstChapter,
            hasTimeSlotAvailable,
            formatWhole,
            openJob,
            unref,
            displayResource,
            resourceArray,
            Visibility,
            Direction
        };
    }
});
</script>

<style scoped>
.job {
    height: 104px;
    width: 516px;
    position: relative;
    background: var(--raised-background);
    overflow: hidden;
    box-shadow: 0 4px 4px 0 rgb(0 0 0 / 25%);
}

.job:not(.selected) {
    border-bottom-left-radius: 50px;
    border-top-left-radius: 50px;
    cursor: pointer;
}

.job:not(.selected):hover {
    box-shadow: 0 0 12px var(--foreground);
}

.job > img {
    position: absolute;
    top: 0;
    left: 0;
    clip-path: circle(100% at 50% 50%);
    z-index: 1;
}

.job:not(.selected),
.job:not(.selected) > * {
    transition: all 0.5s 0.25s, box-shadow 0.5s 0s, clip-path 0s 0s;
}

.job:not(.selected) > img {
    clip-path: circle(51px at var(--posx) var(--posy));
    transform: translate(calc(-1 * var(--posx) + 50px), calc(-1 * var(--posy) + 50px));
    transition: all 0.5s 0.25s, clip-path 0.5s 0s;
}

.job.selected > img {
    height: 288px;
    width: 512px;
    border-radius: var(--border-radius);
    transition: all 0.5s, clip-path 0.5s 0.25s ease-in-out;
}

.job.active {
    border-color: var(--bought);
}

.job.selected {
    height: 314px;
}

.job-loop-toggle {
    position: absolute;
    top: 0;
    right: 0;
    z-index: 10;
}

.job-loop-toggle button {
    color: var(--background);
    border: none;
    padding: 0;
    margin: 10px;
    background: var(--foreground);
    border-radius: 50%;
}

.job-loop-toggle button.active {
    color: var(--link);
    box-shadow: 0 0 8px 4px var(--link);
}

.job-loop-toggle button:not([disabled]):hover {
    box-shadow: 0 0 12px var(--points);
    cursor: pointer;
}

.job-contents {
    position: absolute;
    top: calc(100% - 30px);
    left: 110px;
    right: 10px;
    transform: translateY(-100%);
    z-index: 1;
    text-align: left;
}

.job.selected > .job-contents {
    left: 10px;
}

.job-title {
    margin: 0;
    color: var(--foreground);
    width: 100%;
    text-align: left;
    display: flex;
}

.job-title h2 {
    margin-right: 10px;
}

.job-resource {
    color: var(--foreground);
    display: inline-block;
}

.job-resource + .job-resource::before {
    content: ", ";
}

.job-progress-container {
    position: absolute;
    left: -2px;
    right: -2px;
    bottom: -2px;
    border: 2px solid var(--foreground);
    border-bottom-right-radius: var(--border-radius);
    height: 20px;
}

.job.selected > .job-progress-container {
    left: -2px;
    border-bottom-left-radius: var(--border-radius);
}

.job-progress {
    width: 100%;
    height: 100%;
    background-color: var(--foreground);
    clip-path: inset(0% calc(var(--progress) * (0.88) + 0px) 0% 0%);
    transition: clip-path 0s 0s;
}

.job.selected .job-progress {
    clip-path: inset(0% calc(var(--progress)) 0% 0%);
}

.animating .job-progress {
    transition: clip-path 0.2s 0s;
}

.job-quip {
    position: absolute;
    z-index: 5;
    background: var(--locked);
    border: 2px solid rgba(0, 0, 0, 0.125);
    padding: 5px;
    top: 20px;
    left: 50%;
    transform: translate(-50%);
    color: var(--feature-foreground);
    max-width: 400px;
    font-style: italic;
    pointer-events: none;
    animation: quipAnimation 15s;
    animation-fill-mode: forwards;
}

@keyframes quipAnimation {
    0% {
        opacity: 0;
    }
    5% {
        opacity: 1;
    }
    80% {
        opacity: 1;
    }
    100% {
        opacity: 0;
    }
}
</style>
