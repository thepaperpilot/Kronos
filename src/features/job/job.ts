import {
    Component,
    OptionsFunc,
    GatherProps,
    getUniqueID,
    Replace,
    setDefault,
    StyleValue,
    Visibility,
    CoercableComponent
} from "features/feature";
import JobComponent from "features/job/Job.vue";
import { createResource, Resource } from "features/resources/resource";
import { Persistent, persistent } from "game/persistence";
import Decimal, { DecimalSource } from "util/bignum";
import {
    Computable,
    GetComputableType,
    GetComputableTypeWithDefault,
    processComputable,
    ProcessedComputable
} from "util/computed";
import { createLazyProxy } from "util/proxies";
import { computed, nextTick, ref, Ref, unref } from "vue";
import type { ToastID } from "vue-toastification/dist/types/types";

export const JobType = Symbol("Job");
const levelSoftcapPower = 0.7643; // chosen so that e308 = level 100

declare module "@vue/runtime-dom" {
    interface CSSProperties {
        "--posx"?: string;
        "--posy"?: string;
        "--progress"?: string;
    }
}

export interface JobOptions {
    visibility?: Computable<Visibility>;
    classes?: Computable<Record<string, boolean>>;
    style?: Computable<StyleValue>;
    color: Computable<string>;
    image: Computable<string>;
    imageFocus: Computable<{ x: string; y: string }>;
    resource?: Resource | Resource[];
    randomQuips?: Computable<string[]>;
    layerID: string;
    modifierInfo?: Computable<CoercableComponent>;
    modifierModalAttrs?: Record<string, unknown>;
}

export interface BaseJob {
    id: string;
    name: string;
    xp: Resource;
    rawLevel: Ref<DecimalSource>;
    level: Resource;
    levelProgress: Ref<number>;
    timeLoopActive: Persistent<boolean>;
    currentQuip: Ref<string | null>;
    setQuip: (quip?: string) => void;
    notif?: ToastID;
    type: typeof JobType;
    [Component]: typeof JobComponent;
    [GatherProps]: () => Record<string, unknown>;
}

export type Job<T extends JobOptions> = Replace<
    T & BaseJob,
    {
        visibility: GetComputableTypeWithDefault<T["visibility"], Visibility.Visible>;
        classes: GetComputableType<T["classes"]>;
        style: GetComputableType<T["style"]>;
        color: GetComputableType<T["color"]>;
        image: GetComputableType<T["image"]>;
        imageFocus: GetComputableType<T["imageFocus"]>;
        randomQuips: GetComputableType<T["randomQuips"]>;
    }
>;

export type GenericJob = Replace<
    Job<JobOptions>,
    {
        visibility: ProcessedComputable<Visibility>;
    }
>;

export function createJob<T extends JobOptions>(
    name: string,
    optionsFunc: OptionsFunc<T, BaseJob, GenericJob>
): Job<T> {
    const xp = createResource<DecimalSource>(0, name + " XP");
    const timeLoopActive = persistent<boolean>(false);
    return createLazyProxy(() => {
        const job = optionsFunc();
        job.id = getUniqueID("job-");
        job.type = JobType;
        job[Component] = JobComponent;

        job.name = name;
        job.xp = xp;
        job.timeLoopActive = timeLoopActive;
        job.rawLevel = computed(() => {
            const genericJob = job as GenericJob;
            if (Decimal.eq(genericJob.xp.value, 0)) {
                return 1;
            }
            let baseLevel = Decimal.clampMin(genericJob.xp.value, 1).log10().add(1);
            if (baseLevel.gt(25)) {
                baseLevel = baseLevel.sub(25).pow(levelSoftcapPower).add(25);
            }
            return baseLevel.floor();
        });
        job.level = createResource<DecimalSource>(
            computed(() => (job as GenericJob).rawLevel.value),
            job.name + " Levels"
        );
        job.levelProgress = computed(() => {
            const genericJob = job as GenericJob;
            const level = genericJob.rawLevel.value;

            const prevRequirement = getXPRequirement(level);
            const nextRequirement = getXPRequirement(Decimal.add(level, 1));

            let progress;
            if (Decimal.lt(level, 25)) {
                // Linear
                progress = Decimal.sub(genericJob.xp.value, prevRequirement)
                    .div(Decimal.sub(nextRequirement, prevRequirement))
                    .toNumber();
            } else {
                // Logarithmic
                progress = Decimal.log10(genericJob.xp.value)
                    .sub(Decimal.log10(prevRequirement))
                    .div(Decimal.log10(nextRequirement).sub(Decimal.log10(prevRequirement)))
                    .toNumber();
            }
            return progress;
        });
        job.currentQuip = ref(null);

        job.setQuip = function (quip?: string) {
            const genericJob = job as GenericJob;
            if (genericJob.currentQuip.value) {
                genericJob.currentQuip.value = null;
                nextTick(genericJob.setQuip);
                return;
            }

            if (quip) {
                genericJob.currentQuip.value = quip;
            } else if (unref(genericJob.randomQuips)) {
                const quips = unref(genericJob.randomQuips) as string[];
                genericJob.currentQuip.value = quips[Math.floor(Math.random() * quips.length)];
            }
        };

        processComputable(job as T, "visibility");
        setDefault(job, "visibility", Visibility.Visible);
        processComputable(job as T, "classes");
        processComputable(job as T, "style");
        processComputable(job as T, "color");
        processComputable(job as T, "image");
        processComputable(job as T, "imageFocus");
        processComputable(job as T, "randomQuips");
        processComputable(job as T, "modifierInfo");

        job[GatherProps] = function (this: GenericJob) {
            const {
                id,
                xp,
                level,
                levelProgress,
                timeLoopActive,
                name,
                visibility,
                color,
                image,
                imageFocus,
                resource,
                layerID,
                classes,
                style,
                currentQuip,
                randomQuips,
                modifierInfo,
                modifierModalAttrs
            } = this;
            return {
                id,
                xp,
                level,
                levelProgress,
                timeLoopActive,
                name,
                visibility,
                color,
                image,
                imageFocus,
                resource,
                layerID,
                classes,
                style: unref(style),
                currentQuip,
                randomQuips,
                modifierInfo,
                modifierModalAttrs
            };
        };

        return job as unknown as Job<T>;
    });
}

function getXPRequirement(level: DecimalSource): DecimalSource {
    if (Decimal.eq(level, 1)) {
        return 0;
    }
    if (Decimal.gt(level, 25)) {
        level = Decimal.sub(level, 25)
            .pow(1 / levelSoftcapPower)
            .add(25);
    }
    return Decimal.sub(level, 1).pow10();
}
