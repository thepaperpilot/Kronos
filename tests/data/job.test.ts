import { createJob, GenericJob } from "features/job/job";
import Decimal from "util/bignum";
import { beforeEach, describe, expect, test } from "vitest";
import "../utils";

describe("Level Formula", () => {
    let job: GenericJob;
    beforeEach(() => {
        job = createJob("test", () => ({
            color: "red",
            image: "dummy",
            imageFocus: { x: "0", y: "0" },
            layerID: "dummy",
            symbol: "d"
        }));
    });

    test("Level = 10^n EXP up to 25", () => {
        test.each(new Array(24).fill(0).map((_, i) => i + 1))("%i:", level => {
            job.xp.value = Decimal.pow10(level - 1);
            expect(job.level.value).compare_tolerance(level);
        });
    });

    describe("5.5*10^n is 50% to next level up to 25", () => {
        test.each(new Array(24).fill(0).map((_, i) => i + 1))("%i:", level => {
            job.xp.value = Decimal.pow10(level - 1).times(5.5);
            expect(job.levelProgress.value).compare_tolerance(0.5);
        });
    });

    describe("Level softcap applies after 25", () => {
        describe.each([...new Array(10).fill(0).map((_, i) => i + 25), 100, 1e308])(
            "%i:",
            level => {
                test("XP required", () => expect(job.levelFormula.invert(level)).toMatchSnapshot());
                test("Calculates level correctly", () => {
                    const requiredXp = job.levelFormula.invert(level);
                    job.xp.value = requiredXp;
                    expect(job.level.value).compare_tolerance(level);
                    expect(job.levelProgress.value).compare_tolerance(0);
                });
                // Not sure how to implement this without just duplicating the code itself
                test.todo("Calculates 50% level progress correctly");
            }
        );
    });
});
