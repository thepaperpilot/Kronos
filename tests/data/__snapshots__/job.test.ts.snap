// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Level Formula > Level softcap applies after 25 > 1: > XP required 1`] = `"ee402.9831218108071"`;

exports[`Level Formula > Level softcap applies after 25 > 25: > XP required 1`] = `"1e24"`;

exports[`Level Formula > Level softcap applies after 25 > 26: > XP required 1`] = `"1e25"`;

exports[`Level Formula > Level softcap applies after 25 > 27: > XP required 1`] = `"2.9967080466297658e26"`;

exports[`Level Formula > Level softcap applies after 25 > 28: > XP required 1`] = `"1.6209790383568656e28"`;

exports[`Level Formula > Level softcap applies after 25 > 29: > XP required 1`] = `"1.360716504380679e30"`;

exports[`Level Formula > Level softcap applies after 25 > 30: > XP required 1`] = `"1.6345656220869154e32"`;

exports[`Level Formula > Level softcap applies after 25 > 31: > XP required 1`] = `"2.667606476594593e34"`;

exports[`Level Formula > Level softcap applies after 25 > 32: > XP required 1`] = `"5.701864508501798e36"`;

exports[`Level Formula > Level softcap applies after 25 > 33: > XP required 1`] = `"1.5529644564242793e39"`;

exports[`Level Formula > Level softcap applies after 25 > 34: > XP required 1`] = `"5.275039883599332e41"`;

exports[`Level Formula > Level softcap applies after 25 > 100: > XP required 1`] = `"9.806005514434617e307"`;
